contracts/strategies/StrategyDAI3poolIron.sol

```diff
@@ -0,0 +1,562 @@
+        require(cToken != address(_asset), "cyY3CRV");
+        balance = _asset.balanceOf(address(this));
+        _asset.safeTransfer(controller, balance);
+    }
+
+    // Withdraw partial funds, normally used with a vault withdrawal
+    function withdraw(uint256 _amount) external {
+        require(msg.sender == controller, "!controller");
+
+        rebalance();
+        uint256 _balance = IERC20(want).balanceOf(address(this));
+        if (_balance < _amount) {
+            _amount = _withdrawSome(_amount.sub(_balance));
+            _amount = _amount.add(_balance);
+            tank = 0;
+        } else {
+            if (tank >= _amount) tank = tank.sub(_amount);
+            else tank = 0;
+        }
+
+        address _vault = Controller(controller).vaults(address(want));
+        require(_vault != address(0), "!vault"); // additional protection so we don't burn the funds
+        uint256 _fee = _amount.mul(withdrawalFee).div(DENOMINATOR);
+        IERC20(want).safeTransfer(Controller(controller).rewards(), _fee);
+        IERC20(want).safeTransfer(_vault, _amount.sub(_fee));
+    }
+
+    function _withdrawSome(uint256 _amount) internal returns (uint256) {
+        uint256 _amnt = _amount.mul(1e18).div(ICurveFi(_3pool).get_virtual_price());
+        uint256 _amt = _amnt.mul(1e18).div(yvERC20(y3crv).getPricePerFullShare());
+        uint256 _bal = ICErc20(cToken).balanceOfUnderlying(address(this));
+        if (_amt > _bal) _amt = _bal;
+        require(ICErc20(cToken).redeemUnderlying(_amt) == 0, "!withdrawSome");
+        urn = _bal.sub(_amt);
+        uint256 _before = IERC20(_3crv).balanceOf(address(this));
+        yvERC20(y3crv).withdraw(_amt);
+        uint256 _after = IERC20(_3crv).balanceOf(address(this));
+        return _withdrawOne(_after.sub(_before));
+    }
+
+    function _withdrawOne(uint256 _amnt) internal returns (uint256) {
+        uint256 _before = IERC20(want).balanceOf(address(this));
+        IERC20(_3crv).safeApprove(_3pool, 0);
+        IERC20(_3crv).safeApprove(_3pool, _amnt);
+        ICurveFi(_3pool).remove_liquidity_one_coin(_amnt, 0, _amnt.mul(DENOMINATOR.sub(slip)).div(DENOMINATOR));
+        uint256 _after = IERC20(want).balanceOf(address(this));
+
+        return _after.sub(_before);
+    }
+
+    // Withdraw all funds, normally used when migrating strategies
+    function withdrawAll() external returns (uint256 balance) {
+        require(msg.sender == controller, "!controller");
+        _withdrawAll();
+
+        balance = IERC20(want).balanceOf(address(this));
+
+        address _vault = Controller(controller).vaults(address(want));
+        require(_vault != address(0), "!vault"); // additional protection so we don't burn the funds
+        IERC20(want).safeTransfer(_vault, balance);
+    }
+
+    function _withdrawAll() internal {
+        uint256 _y3crv = ICErc20(cToken).balanceOfUnderlying(address(this));
+        require(ICErc20(cToken).redeemUnderlying(_y3crv) == 0, "!withdrawAll");
```

Comment:
is it possible that `_y3crv` is zero at any point in time? and if that so can `redeemUnderlying` return error (non-zero) response if we pass 0 as argument?

Reply:
yes, it is possible. but `redeemUnderlying` with 0 amount returns 0, which is successful at command level, just you don get any underlying y3crv.